lib/snippets/kubernetes-deploy

```diff
@@ -66,99 +200,455 @@ class KubernetesDeploy
+      @containers = pod_data["spec"]["containers"].map { |c| c["name"] }
+
+      if @deploy_started && pod_data["status"]["containerStatuses"]
+        pod_data["status"]["containerStatuses"].each do |status|
+          reason = status.dig("state", "waiting", "reason")
+          next unless SUSPICIOUS_CONTAINER_STATES.include?(reason)
+          message = status.dig("state", "waiting", "message")
+          KubernetesDeploy.logger.warn("#{id} has container in state #{reason} (#{message})")
+        end
+      end
+
+      if @phase == "Failed"
+        @status = "#{@phase} (Reason: #{pod_data["status"]["reason"]})"
+      elsif @phase == "Terminating"
+        @status = @phase
+      else
+        ready_condition = pod_data["status"]["conditions"].find { |condition| condition["type"] == "Ready" }
+        @ready = ready_condition ? (ready_condition["status"] == "True") : false
+        @status = "#{@phase} (Ready: #{@ready})"
+      end
+    end
+
+    def deploy_succeeded?
+      if @bare
+        @phase == "Succeeded"
+      else
+        @phase == "Running" && @ready
+      end
+    end
+
+    def deploy_failed?
+      @phase == "Failed"
+    end
+
+    def exists?
+      @bare ? @found : true
+    end
+
+    def group_name
+      @bare ? "Bare pods" : @parent
+    end
+
+    private
+
+    def display_logs
+      return {} unless exists? && @containers.present? && !@already_displayed
+
+      @containers.each do |container_name|
+        out, st = run_kubectl("logs", @name, "--timestamps=true", "--since-time=#{@deploy_started.to_datetime.rfc3339}")
+        next unless st.success? && out.present?
+
+        KubernetesDeploy.logger.info "Logs from #{id} container #{container_name}:"
+        STDOUT.puts "#{out}"
+        @already_displayed = true
+      end
+    end
+  end
+
+  class Deployment < KubernetesResource
+    TIMEOUT = 60 * 15
+
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      json_data, st = run_kubectl("get", type, @name, "--output=json")
+      @found = st.success?
+      @rollout_data = {}
+      @status = nil
+      @pods = []
+
+      if @found
+        @rollout_data = JSON.parse(json_data)["status"].slice("updatedReplicas", "replicas", "availableReplicas", "unavailableReplicas")
+        @status, _ = run_kubectl("rollout", "status", type, @name, "--watch=false") if @deploy_started
+
+        pod_list, st = run_kubectl("get", "pods", "-a", "-l", "name=#{name}", "--output=json")
+        if st.success?
+          pods_json = JSON.parse(pod_list)["items"]
+          pods_json.each do |pod_json|
+            pod_name = pod_json["metadata"]["name"]
+            pod = Pod.new(pod_name, namespace, nil, parent: "#{@name.capitalize} deployment")
+            pod.deploy_started = @deploy_started
+            pod.interpret_json_data(pod_json)
+            pod.log_status
```

Comment:
The idea behind this logging is that a visualization script will make little boxes for all the resources and make them change colours as they change state. Therefore, it's important to write a line for every resource. It's not meant to be human-readable (though it is somewhat useful for debugging in the meantime).

Note: I discovered that, somewhat unsurprisingly, the overhead for including logs and/or events in this output was unacceptably enormous, so  I ripped it all out; this logging no longer triggers any additional requests. After all, logs are available in splunk, and events _should_ be (but aren't iirc). A potential future enhancement could be to fetch and dump logs and events for failed resources only at the time a deploy fails.